### PR TITLE
Validate target references in custom workspace schemes

### LINKED ISSUE #6510: Targets referenced from a workspace scheme build step aren't validated
### What problem or need do you have?

When I raised the issue of custom, non-Tuist defined schemes not being supported (https://github.com/tuist/tuist/issues/6480), I was advised to define schemes in Workspace.swift instead of Xcode.

While experimenting with this, I noticed that if I want a scheme to have a build action, the target that is supposed to be built is not validated for existence, and Tuist produces an empty build action if a target is invalid.

In other words, given this Workspace.swift:

```
import ProjectDescription

let workspace = Workspace(
    name: "MyWorkspace",
    projects: [
        "MyProject",
    ],
    schemes: [
        .scheme(
            name: "MyScheme",
            buildAction: .buildAction(
                targets: [.target("BASICALLY ANY TEXT WORKS")]
            ),
            testAction: .testPlans([
               ...
            ]),
            runAction: .runAction(
                ...
            )
        ),
    ]
)

```

I can type any text into instead of 'BASICALLY ANY TEXT WORKS' and Tuist will happily generate the scheme. If the target with this name doesn't exist, Tuist doesn't output any warnings or errors, and this is how the resulting scheme's build action looks like:

```
<BuildAction
      parallelizeBuildables = "YES"
      buildImplicitDependencies = "YES">
<BuildActionEntries></BuildActionEntries>
</BuildAction>
```

### Potential solution

There is some validation happening at the workspace level (for example, the projects for the `projects` property are checked for existence and there's a warning if a project doesn't exist).

Validation could be extended to the targets referenced from the workspace. I don't know the order of the things happening, but looking at the output it appears that the Workspace is generated before the projects (and their targets), and checking for the projects's existence doesn't require the project to be generated (hence the existing validation that works). If true, this means the proper way to solve this would be to generate the projects before the workspace, unless there's any reason not to do this – and then apply the knowledge of the projects and their constituent targets during the validation of Workspace.swift.

### macOS version

14.5

### Tuist version

4.20.0

### Xcode version

15.4

### PR DESCRIPTION
**Pull request by ajkolean**
Resolves #6510

### Short description 📝

This PR enhances the `ManifestLinter` to validate target references in workspace schemes, ensuring that targets are correctly specified with their project paths. This improvement prevents the generation of invalid build actions in custom, non-Tuist defined schemes.

### Key features:

- **Target Reference Validation**: Ensures that targets in workspace schemes specify their project paths using `.project(path:, target:)`.
- **Comprehensive Linting**: Adds linting for `buildAction`, `runAction`, and `profileAction` in workspace schemes to check for unspecified `projectPaths`.
- **User-Friendly Warnings**: Provides clear and friendly error messages, indicating the need to specify project paths and highlighting the source of the issue in `Workspace.swift`.
- **Improved Scheme Integrity**: Prevents the generation of empty build actions by ensuring all targets are validated for existence and correctness.

### How to test the changes locally 🧐

1. **Prepare Workspace.swift**:
    - Define schemes in `Workspace.swift` with targets that lack project paths.
    - Example:
      ```swift
      import ProjectDescription

      let workspace = Workspace(
          name: "MyWorkspace",
          projects: [
              "MyProject",
          ],
          schemes: [
              .scheme(
                  name: "MyScheme",
                  buildAction: .buildAction(
                      targets: [.target("TargetA")]
                  ),
                  runAction: .runAction(
                      expandVariableFromTarget: .target("TargetA")
                  ),
                  profileAction: .profileAction(
                      executable: .target("TargetA")
                  )
              ),
          ]
      )
      ```

2. **Run Linter**:
    - Execute the linter and observe the warnings:
      ```sh
      swift run tuist generate --path path/to/project --no-open 
      ```

3. **Verify Warnings**:
    - Confirm that warnings are displayed for each instance where a target is missing its project path:
      ```
      Workspace.swift: The target 'InvalidTarget' in the scheme 'MyScheme' is missing the project path.
      Please specify the project path using .project(path:, target:).
      ```

### Contributor checklist ✅

- [x] The code has been linted using `run mise run lint:fix`.
- [x] The change is tested via unit testing or acceptance testing, or both.
- [x] The title of the PR is formulated in a way that is usable as a changelog entry.
- [x] In case the PR introduces changes that affect users, the documentation has been updated.

### Reviewer checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase.
- [x] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry.

### CODE REVIEW

File: Sources/TuistKit/Utils/ManifestGraphLoader.swift [hunk 1]
```diff
@@ -143,8 +143,10 @@ public final class ManifestGraphLoader: ManifestGraphLoading {
         )
 
         // Lint Manifests
-        let lintingIssues = manifestProjects.flatMap { manifestLinter.lint(project: $0.value) }
-        try lintingIssues.printAndThrowErrorsIfNeeded()
+        let workspaceLintingIssues = manifestLinter.lint(workspace: allManifests.workspace)
+        let projectLintingIssues = manifestProjects.flatMap { manifestLinter.lint(project: $0.value) }
+        try projectLintingIssues.printAndThrowErrorsIfNeeded()
+        let lintingIssues = workspaceLintingIssues + projectLintingIssues
```
Comment (kwridan) [inline Sources/TuistKit/Utils/ManifestGraphLoader.swift:149]: The `printAndThrowErrorsIfNeeded()` method stops the generation process by throwing in the event lint issues with severity `error` are encountered. 

We can print and throw the combined linting issue as the workspace linter may also produce error lint issues.

```suggestion
        let lintingIssues = workspaceLintingIssues + projectLintingIssues
        try lintingIssues.printAndThrowErrorsIfNeeded()

```

File: Sources/TuistLoader/Linter/ManifestLinter.swift [hunk 2]
```diff
@@ -39,6 +53,83 @@ public class ManifestLinter: ManifestLinting {
...
+            if let buildAction = scheme.buildAction {
+                issues.append(contentsOf: lintWorkspaceSchemeTargets(
+                    buildAction.targets,
+                    actionType: "buildAction",
+                    scheme: scheme
+                ))
+            }
+
+            if let runAction = scheme.runAction {
+                issues.append(contentsOf: lintWorkspaceSchemeTarget(
```
Comment (kwridan) [inline Sources/TuistLoader/Linter/ManifestLinter.swift:69]: The runAction can specify an executable:

https://github.com/tuist/tuist/blob/main/Sources/ProjectDescription/RunAction.swift#L23

**Review by kwridan — COMMENTED** [review]
Thanks for working on this @ajkolean 🙏

There quite a few other `TargetReference`s throughout the scheme API that can also benefit from being included:

- [`TestAction.targets`](https://github.com/tuist/tuist/blob/main/Sources/ProjectDescription/TestAction.swift#L12) contain target references that can benefit from the same check 
- All the `pre` / `post` [ExecuteAction](https://github.com/tuist/tuist/blob/main/Sources/ProjectDescription/ExecuteAction.swift#L7)s in all the scheme actions also contain target references

**Comment by ajkolean** [comment]
> Thanks for working on this @ajkolean 🙏
> 
> There quite a few other `TargetReference`s throughout the scheme API that can also benefit from being included:
> 
> * [`TestAction.targets`](https://github.com/tuist/tuist/blob/main/Sources/ProjectDescription/TestAction.swift#L12) contain target references that can benefit from the same check
> * All the `pre` / `post` [ExecuteAction](https://github.com/tuist/tuist/blob/main/Sources/ProjectDescription/ExecuteAction.swift#L7)s in all the scheme actions also contain target references

@kwridan Good callout! Updated to handle these other case.

**Review by kwridan — APPROVED** [review]
awesome, thank you for the updates @ajkolean

**Review by fortmarek — APPROVED** [review]
Thanks 👏 

The alternative would to change the `schemes: [Scheme]` property of `Workspace.swift` to `schemes: [WorkspaceScheme]` but we'd end up up with a lot of duplication – and it would be a breaking change.

LGTM, long-term, we can consider if the having a dedicated `WorkspaceScheme` would make sense to ensure the validity of actions during the manifest's compile time instead of during the linting phase.

**Pull request merged**